src/protocol/codecs/topic.rs

```diff
@@ -0,0 +1,113 @@
+use std::io::{self, ErrorKind};
+
+use bytes::BytesMut;
+use tokio_util::codec::Decoder;
+
+use crate::protocol::invalid_data;
+
+/// Topic is a key-value string pair.
+pub struct Topic {
+    /// Key.
+    pub key: String,
+
+    /// Value.
+    pub value: String,
+}
+
+#[derive(Default, Clone)]
+pub struct TopicCodec;
+
+impl Decoder for TopicCodec {
+    type Item = Vec<Topic>;
+    type Error = io::Error;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        if src.is_empty() {
+            return Ok(Some(vec![]));
+        }
+
+        let (&num_topics, mut src) = src
+            .split_first()
```

Comment:
if we used some [`Buf ](https://docs.rs/bytes/latest/bytes/buf/trait.Buf.html)`method here instead, we wouldn't need to clone some of those first items later on